From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gabriel Ravier <gabravier@gmail.com>
Date: Tue, 11 May 2021 19:43:43 +0200
Subject: [PATCH] 3026: Add call fixup for edi on IA-32 and a corresponding
 pattern

Added call fixup on IA-32 for get_pc_thunk_di and a corresponding pattern

Also fixed formatting on the line with `callfixup name="get_pc_thunk_si"`
---
 Ghidra/Processors/x86/data/languages/x86gcc.cspec   | 13 ++++++++++++-
 .../x86/data/patterns/x86gcc_patterns.xml           |  5 +++++
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/Ghidra/Processors/x86/data/languages/x86gcc.cspec b/Ghidra/Processors/x86/data/languages/x86gcc.cspec
index bcd049d2d..25ede8426 100644
--- a/Ghidra/Processors/x86/data/languages/x86gcc.cspec
+++ b/Ghidra/Processors/x86/data/languages/x86gcc.cspec
@@ -364,7 +364,7 @@
     </pcode>
   </callfixup>
   
-    <callfixup name="get_pc_thunk_si">
+  <callfixup name="get_pc_thunk_si">
     <target name="__i686.get_pc_thunk.si"/>
     <target name="__x86.get_pc_thunk.si"/>
     <pcode>
@@ -374,4 +374,15 @@
       ]]></body>
     </pcode>
   </callfixup>
+
+  <callfixup name="get_pc_thunk_di">
+    <target name="__i686.get_pc_thunk.di"/>
+    <target name="__x86.get_pc_thunk.di"/>
+    <pcode>
+      <body><![CDATA[
+      EDI = * ESP;
+      ESP = ESP + 4;
+      ]]></body>
+    </pcode>
+  </callfixup>
 </compiler_spec>
diff --git a/Ghidra/Processors/x86/data/patterns/x86gcc_patterns.xml b/Ghidra/Processors/x86/data/patterns/x86gcc_patterns.xml
index 89885e5ae..39c19103d 100644
--- a/Ghidra/Processors/x86/data/patterns/x86gcc_patterns.xml
+++ b/Ghidra/Processors/x86/data/patterns/x86gcc_patterns.xml
@@ -90,6 +90,11 @@
      <funcstart label="__i686.get_pc_thunk.si" validcode="function"/>
   </pattern>
 
+  <pattern>
+     <data>0x8b 0x3c 0x24 0xc3 </data> <!-- MOV  EDI,[ESP] / RET -->
+     <funcstart label="__i686.get_pc_thunk.di" validcode="function"/>
+  </pattern>
+
   <patternpairs totalbits="32" postbits="16">
     <prepatterns>
       <data>0x90</data> <!-- NOP filler -->
-- 
2.39.1

